### DISCUSSION TITLE
Windows Installer

### DISCUSSION THREAD

**Discussion by 0Grit** [pyocd/pyOCD#1274] [Ideas]
_This has been discussed a couple times in slack but I don't see an issue or discussion for it yet_

I think the project would see a significant increase in users, (and eventually contributors) if pyOCD had an official windows installer.
https://github.com/flit/pyOCD/tree/feature/binary_build

@flit what is the current blocker of this again?

**Comment by flit** [comment]
Hi @0Grit, really sorry I missed this!

There are two possibilities for an installer:

1. Binary executable build of the pyocd command line tool. Not sure if an installer would even be necessary since it's just one file, I don't know Windows enough to know the answer to that.
2. Packaged installation of Python 3.x + pyocd. The suggestion on Slack was to use [BeeWare Briefcase](https://github.com/beeware/briefcase).

The blocker for the first is that pyInstaller doesn't support package entry points as used by pyocd for its debug probe and RTOS plugins. There are some options other than pyInstaller, eg pyOxidizer, but they haven't been tried yet and therefore have many unknowns.

**Comment by elfmimi** [comment]
I tried **pyoxidizer** and **pyinstaller** . My goal is to reduce launch time for pyocd commander along with use of user script. So I ditched everything not needed for this purpose. No capstone. No cmsis-pack-manager. No built-in targets other than cortex_m . No gdbserver. and Only support CMSIS-DAP as interface.
I didn't want to bother with entrypoint things. So I did this.
```python
from ..probe.cmsis_dap_probe import CMSISDAPProbe
PROBE_CLASSES["cmsisdap"] = CMSISDAPProbe
```
instead of
```python
load_plugin_classes_of_type('pyocd.probe', PROBE_CLASSES, DebugProbe)
```
And using impromptu startup script like this.
```python
from pyocd.__main__ import main
main()
```
With pyinstaller, all I had to do was: `pyinstaller main.py --add-binary="libusb-1.0.dll;." -n pyocd`
Note: I'm using pre libusb-package version.
Somehow pyoxidizer makes things go slower! So I'm not showing its detail here.
Pyinstaller do a better job and actually reduces launch time. 👍 
`pyocd list` : from 0.78 sec to 0.48 sec
`pyocd cmd -t cortex_m --script nuvoton.py -c ident` : from 1.75 sec to 1.35 sec
Note: Yes! Windows is very slow by its nature!
--onefile option functions as it is described. You'll get just one single .exe file. Sadly this one takes longer to launch than plain python. 👎

**Reply by CyberXNomad** [reply]
Hi @elfmimi . it can't probe debuggers after I package pyocd with Pyinstaller on Windows. Do you have any idea on this?

**Reply by elfmimi** [reply]
Hi. What probe are you using? Begin with trying DAPLink or other CMSIS-DAP based probe first.

**Reply by CyberXNomad** [reply]
I have two different CMSIS-DAP based debuggers. They can be listed in pyocd api based script. But can't  be listed after packaged.
```python
# test.py
probe_list = ConnectHelper.get_all_connected_probes(blocking=True)
# success get list by directly call `python test.py`
# but block in this function after packaged and console output `Waiting for a debug probe to be connected...`
```

**Reply by elfmimi** [reply]
Did you change this line? https://github.com/pyocd/pyOCD/blob/88222188de5ee553ada083efbdd54eb8f15974b0/pyocd/probe/aggregator.py#L83

**Reply by CyberXNomad** [reply]
No, I didn't. Should I?

**Reply by elfmimi** [reply]
Yup. See my first comment.

**Reply by elfmimi** [reply]
May be you need to copy libusb-1.0.dll next to your newly built binary file.

**Reply by CyberXNomad** [reply]
>So I ditched everything not needed for this purpose. No capstone. No cmsis-pack-manager. No built-in targets other than cortex_m . No gdbserver. and Only support CMSIS-DAP as interface.

If I change that line, will that remove `cmsis-pack-manager` and `built-in targets`? I thought modifying that line would remove `cmsis-pack-manager` and `built-in targets`, I need both, so I don't think I should modify it. 
I'll try it first.
In fact, my purpose is not to package pyocd, but to package scripts based on pyocd api.

**Reply by CyberXNomad** [reply]
Great! That works. Thank you very much. <3

**Reply by elfmimi** [reply]
Thank you @xray-bit . I've learned something new from your question.
You can do this ↓ with unmodified HEAD of pyOCD.

cd pyOCD

create test.py .
```python
from pyocd.probe.aggregator import PROBE_CLASSES
from pyocd.probe.cmsis_dap_probe import CMSISDAPProbe
PROBE_CLASSES["cmsisdap"] = CMSISDAPProbe

from pyocd.core.helpers import ConnectHelper
probe_list = ConnectHelper.get_all_connected_probes(blocking=True)
print(probe_list)
```
copy libusb-1.0.dll to the current directory.

build it.
`pyinstaller test.py --add-binary="libusb-1.0.dll;." -n pyocd-test`

then execute.
`.\dist\pyocd-test\pyocd-test.exe`

**Reply by CyberXNomad** [reply]
Oh, that's a better way!
I'm a python beginner, so I don't know enough about the code structure of the whole project. 
Thank you for your help again.
